src/main/java/com/puppycrawl/tools/checkstyle/TreeWalker.java

```diff
@@ -144,26 +176,52 @@ else if (module instanceof TreeWalkerFilter) {
         }
     }
 
+    /**
+     * {@inheritDoc} Processes the file.
+     *
+     * @noinspection ProhibitedExceptionThrown
+     * @noinspectionreason ProhibitedExceptionThrown - there is no other way to obey
+     *     skipFileOnJavaParseException field
+     */
     @Override
     protected void processFiltered(File file, FileText fileText) throws CheckstyleException {
         // check if already checked and passed the file
         if (!ordinaryChecks.isEmpty() || !commentChecks.isEmpty()) {
             final FileContents contents = getFileContents();
-            final DetailAST rootAST = JavaParser.parse(contents);
-            if (!ordinaryChecks.isEmpty()) {
-                walk(rootAST, contents, AstState.ORDINARY);
-            }
-            if (!commentChecks.isEmpty()) {
-                final DetailAST astWithComments = JavaParser.appendHiddenCommentNodes(rootAST);
-                walk(astWithComments, contents, AstState.WITH_COMMENTS);
+            DetailAST rootAST = null;
+            // whether skip the procedure after parsing Java files.
+            boolean skip = false;
+            try {
+                rootAST = JavaParser.parse(contents);
             }
-            if (filters.isEmpty()) {
+            // -@cs[IllegalCatch] There is no other way to obey skipFileOnJavaParseException field
+            catch (Exception ex) {
+                if (!skipFileOnJavaParseException) {
+                    throw ex;
+                }
+                skip = true;
+                violations.add(new Violation(1, Definitions.CHECKSTYLE_BUNDLE, PARSE_EXCEPTION_MSG,
+                            new Object[] {ex.getMessage()}, javaParseExceptionSeverity, null,
+                            getClass(), null));
```

Comment:
We can move the parsing and its exception handling part to a new method that returns DetailAST (rootAST)

Reply:
I just sharing my thoughts, this can be improved later but to speed up reviewing and merging let's keep the code in the same method for now.

Reply:
>I just sharing my thoughts, this can be improved later but to speed up reviewing and merging let's keep the code in the same method for now.

I am good with this, this is a long awaited user feature.

Reply:
> exception handling part to a new method that returns DetailAST 

This is basically what `JavaParser` is, and I mentioned before it would make sense for some of this handling to be in there wrapped around the `new JavaAstVisitor`.

It also makes sense to me to not skip the filters when we create a violation. We are basically preventing the users from handling it in TW and forcing them to use something in Checker.

We can make these in another PR. My items below are my last.